x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java

```diff
@@ -64,427 +113,1139 @@
+
+        for (CandidateTimestampFormat candidate : orderedCandidateFormats) {
+
+            TimestampMatch match = checkCandidate(candidate, text, numberPosBitSet, requireFullMatch, timeoutChecker);
+            if (match != null) {
+                TimestampFormat newFormat = match.timestampFormat;
+                boolean mustAdd = true;
+                for (int i = 0; i < matchedFormats.size(); ++i) {
+                    TimestampFormat existingFormat = matchedFormats.get(i);
+                    if (existingFormat.canMergeWith(newFormat)) {
+                        matchedFormats.set(i, existingFormat.mergeWith(newFormat));
+                        mustAdd = false;
                         break;
                     }
                 }
-                if (quicklyRuledOut == false) {
-                    Map<String, Object> captures = timeoutChecker.grokCaptures(candidate.strictSearchGrok, text,
-                        "timestamp format determination");
-                    if (captures != null) {
-                        String preface = captures.getOrDefault(PREFACE, "").toString();
-                        String epilogue = captures.getOrDefault(EPILOGUE, "").toString();
-                        return makeTimestampMatch(candidate, index, preface, text.substring(preface.length(),
-                            text.length() - epilogue.length()), epilogue);
+                if (mustAdd) {
+                    if (errorOnMultiplePatterns && matchedFormats.isEmpty() == false) {
+                        throw new IllegalArgumentException("Multiple timestamp formats found ["
+                            + matchedFormats.get(0) + "] and [" + newFormat + "]");
                     }
+                    matchedFormats.add(newFormat);
                 }
+
+                matches.add(match);
+                cachedJavaTimestampFormats = null;
+                return;
             }
-            ++index;
         }
-        return null;
+
+        if (errorOnNoTimestamp) {
+            throw new IllegalArgumentException("No timestamp found in [" + text + "]");
+        }
     }
 
     /**
-     * Find the best timestamp format for matching an entire field value.
-     * @param text The value that the returned timestamp format must match in its entirety.
-     * @param timeoutChecker Will abort the operation if its timeout is exceeded.
-     * @return The timestamp format, or <code>null</code> if none matches.
+     * Where multiple timestamp formats have been found, select the "best" one, whose details
+     * will then be returned by methods such as {@link #getGrokPatternName} and
+     * {@link #getJavaTimestampFormats}.  If fewer than two timestamp formats have been found
+     * then this method does nothing.
+     */
+    public void selectBestMatch() {
+
+        if (matchedFormats.size() < 2) {
+            // Nothing to do
+            return;
+        }
+
+        int remainingMatches = matches.size();
```

Comment:
This block seems a nice candidate to extract into a method that returns the weights.